### PR TITLE
feat(launchdarkly-provider): Add provider

### PR DESCRIPTION
**Pull request by c4milo**
This PR implements an OpenFeature provider for LaunchDarkly, it includes tests and 1 example.

### CODE REVIEW

**Comment by Kavindu-Dodan** [comment]
@c4milo  thank you for the contribution 🙌 

We will review the contribution and meanwhile, could you please sign off the commits? [1]. You can do this with a rebase for your commits with `--signoff` flag [2] 

[1] - https://github.com/open-feature/go-sdk-contrib/pull/178/checks?check_run_id=12881340642
[2] - https://git-scm.com/docs/git-rebase#Documentation/git-rebase.txt---signoff

File: providers/launchdarkly/go.mod [hunk 1]
```diff
@@ -0,0 +1,29 @@
+module github.com/open-feature/go-sdk-contrib/providers/launchdarkly
+
+go 1.20
```
Comment (Kavindu-Dodan) [inline providers/launchdarkly/go.mod:3]: This needs to be adjusted to comply with LD supported version - which is 1.18 [1][2]

[1] - https://github.com/launchdarkly/go-server-sdk#supported-go-versions
[2] - https://github.com/launchdarkly/go-sdk-common#supported-go-versions

File: providers/launchdarkly/pkg/provider.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+type Option func(*options)
+
+// options contains all the optional argumnents supported by Provider.
+type options struct {
+	kindAttr string
+	l        Logger
+}
+
+// WithLoggers sets a logger implementation. By default a noop logger is used.
```
Comment (Kavindu-Dodan) [inline providers/launchdarkly/pkg/provider.go:32]: lint - `WithLogger`

**Comment by Kavindu-Dodan** [comment]
@c4milo I did an initial check and commented some observations.

Besides, you also need to add missing release please configurations. You can add this easily by running the make command [1] and optionally provide release please manifest entry


[1] - https://github.com/open-feature/go-sdk-contrib/blob/main/CONTRIBUTING.md#adding-a-module

File: providers/launchdarkly/pkg/provider.go (hunk 2 above)
Comment (c4milo) [reply providers/launchdarkly/pkg/provider.go:32]: ah, good catch!

File: providers/launchdarkly/pkg/provider.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+		ldCtx.Anonymous(anon)
+	}
+
+	if privateAttrs, ok := evalCtx["privateAttributes"].([]string); ok {
+		ldCtx.Private(privateAttrs...)
+	}
+
+	for key, value := range evalCtx {
+		// skip attributes that were already added to the context
+		if slices.Contains([]string{
```
[the remark this answers is not in the document]
Comment (c4milo) [reply providers/launchdarkly/pkg/provider.go:126]: yes, nice. Thanks for the recommendation.

**Comment by c4milo** [comment]
@Kavindu-Dodan, thanks for the great feedback. I think this is ready for another round of reviews. I'm not too familiar with Release Please but I did review it and it seems this PRs commits is meeting it but I'm not too sure about it.

File: providers/launchdarkly/pkg/provider.go [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+	if kind == ldcontext.MultiKind {
+		p.l.Debug("multi context detected")
+		return p.toMultiLDContext(evalCtx)
+	}
+
+	p.l.Debug("single context detected")
+	return p.mapContext(kind, evalCtx)
+}
+
+// toReason maps LaunchDarkly flag evalution reasons to OpenFeatures'
```
Comment (Kavindu-Dodan) [inline providers/launchdarkly/pkg/provider.go:167]:
```suggestion
// toReason maps LaunchDarkly flag evaluation reasons to OpenFeatures'
```

File: providers/launchdarkly/pkg/provider.go [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+)
+
+var errKeyMissing = errors.New("key and targetingKey attributes are missing, at least 1 required")
+
+// Scream at compile time if Provider does not implement FeatureProvider
+var _ openfeature.FeatureProvider = (*Provider)(nil)
+
+type Option func(*options)
+
+// options contains all the optional argumnents supported by Provider.
```
Comment (Kavindu-Dodan) [inline providers/launchdarkly/pkg/provider.go:26]:
```suggestion
// options contains all the optional arguments supported by Provider.
```

**Comment by Kavindu-Dodan** [comment]
> @Kavindu-Dodan, thanks for the great feedback. I think this is ready for another round of reviews. I'm not too familiar with Release Please but I did review it and it seems this PR's commits is meeting it but I'm not too sure about it.

Cool. I left some language reviews based on my IDE feedback. 

Regarding release please, you can refer this example and add a manual entry to `release-please-config.json` [1]. That's all you need to do and this will make releases to be packaged automatically. 

[1] - https://github.com/open-feature/go-sdk-contrib/blob/main/release-please-config.json#L23-L29

**Review by Kavindu-Dodan — APPROVED** [review]
Approving from my end & overl looks good 🙌

**Comment by c4milo** [comment]
@Kavindu-Dodan, thanks for your kind review and time! much appreciated. Please let me know what else is needed to get it merged, I would like to point my team to the `open-feature/go-sdk-contrib` repo instead of my personal fork.

**Comment by toddbaert** [comment]
@c4milo I will look at this myself in the next day or so. It would also be great to have a review from somebody at LaunchDarkly, if possible. Paging @InTheCloudDan and @Kavindu-Dodan , maybe they can take a look or refer someone else who can.

**Comment by c4milo** [comment]
@toddbaert, thank you! I will reach out also to LaunchDarkly from our end to see if they can review.

File: providers/launchdarkly/pkg/provider.go [hunk 6]
```diff
@@ -0,0 +1,366 @@
...
+
+	return ofDetail
+}
+
+// resolveContext encapsulates common logic that validates the OpenFeature evaluation
+// contest and translates it to LaunchDarkly's. Callers are supposed to only use
+// ldCtx if an error is not returned. When an error is returned, a non-empty OpenFeature
+// ResolutionDetail is returned as well with additional details.
+// This function also handles context cancellations.
+func (p *Provider) resolveContext(ctx context.Context, evalCtx openfeature.FlattenedContext) (ldcontext.Context, openfeature.InterfaceResolutionDetail, error) {
```
Comment (toddbaert) [inline providers/launchdarkly/pkg/provider.go:228]: [`Resolution`](https://openfeature.dev/specification/glossary#resolving-flag-values) is the term we use to mean "get flag value from provider". I think `tranformContext` might be a bit better a name for this func.

File: providers/launchdarkly/pkg/provider.go [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+	ldCtx, err := p.toLDContext(evalCtx)
+	emptyDetail := openfeature.InterfaceResolutionDetail{}
+
+	if err != nil {
+		errMsg := err.Error()
+		detail := openfeature.InterfaceResolutionDetail{
+			ProviderResolutionDetail: openfeature.ProviderResolutionDetail{
+				ResolutionError: openfeature.NewInvalidContextResolutionError(errMsg),
+				Reason:          openfeature.DefaultReason,
```
Comment (toddbaert) [inline providers/launchdarkly/pkg/provider.go:237]: In error scenarios, the `Reason` should indicate an error: `openfeature.ErrorReason`. You can also add an informative message in the `ResolutionError` member.

File: providers/launchdarkly/pkg/provider.go (hunk 6 above)
Comment (c4milo) [reply providers/launchdarkly/pkg/provider.go]: ah, that makes sense! Thank @toddbaert!

File: providers/launchdarkly/pkg/provider.go (hunk 7 above)
Comment (c4milo) [reply providers/launchdarkly/pkg/provider.go]: I see, fair 🙌🏻

**Review by kinyoklion — COMMENTED** [review]
I've done an initial look through the code, it all looks reasonable to me and I don't see anything that would cause any problems on the LaunchDarkly side of things. Great work @c4milo!

I appreciate the way the structure mirrors the other OpenFeature providers we have authored.

Thanks for your patience as well. I should be able to provide some additional feedback or approval soon.

File: providers/launchdarkly/README.md [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# LaunchDarkly OpenFeature Provider for Go
```
[the remark this answers is not in the document]
Comment (c4milo) [reply providers/launchdarkly/README.md:1]: Sounds reasonable!

**Review by kinyoklion — APPROVED** [review]

File: providers/launchdarkly/README.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (toddbaert) [reply providers/launchdarkly/README.md]: @kinyoklion when we add this to our doc-page, we will not put the "official" check on it, and will denote that it's community maintained in the title.

**Comment by Kavindu-Dodan** [comment]
Thank you again for the contribution @c4milo. I will merge this with the approval from the community as well as LD side

**Comment by c4milo** [comment]
Thank you @Kavindu-Dodan!

**Pull request merged**